chainstore.go

```diff
@@ -1,147 +1,228 @@
+	defer s.errMu.Unlock()
+	s.errE = err
+}
+
+// Chain represents a store chain.
 type Chain struct {
-	stores []Store
-	async  bool
+	stores []*storeWrapper
 }
 
+// New creates a new store chain backed by the passed stores.
 func New(stores ...Store) Store {
-	return &Chain{stores, false}
-	// TODO: make the chain..
-	// call Open(), but in case of error..?
+	c := &Chain{
+		stores: make([]*storeWrapper, 0, len(stores)),
+	}
+	for _, s := range stores {
+		c.stores = append(c.stores, &storeWrapper{Store: s})
+	}
+	return c
 }
 
-func Async(stores ...Store) Store {
-	return &Chain{stores, true}
-}
+// Open opens all the stores.
+func (c *Chain) Open() error {
 
-func (c *Chain) Open() (err error) {
-	for _, s := range c.stores {
-		err = s.Open()
-		if err != nil {
-			return // return first error that comes up
-		}
+	if err := c.firstErr(); err != nil {
+		return fmt.Errorf("Open failed due to a previous error: %q", err)
 	}
-	return
+
+	var wg sync.WaitGroup
+
+	for i := range c.stores {
+		wg.Add(1)
+		go func(s *storeWrapper) {
+			defer wg.Done()
+			s.setErr(s.Open())
+		}(c.stores[i])
+	}
+
+	wg.Wait()
+
+	return c.firstErr()
 }
 
-func (c *Chain) Close() (err error) {
-	for _, s := range c.stores {
-		err = s.Close()
-		// TODO: we shouldn't stop on first error.. should keep trying to close
-		// and record errors separately
-		if err != nil {
-			return
-		}
+// Close closes all the stores.
+func (c *Chain) Close() error {
+	var wg sync.WaitGroup
+
+	for i := range c.stores {
+		wg.Add(1)
+		go func(s *storeWrapper) {
+			defer wg.Done()
+			s.setErr(s.Close())
+		}(c.stores[i])
 	}
-	return
+
+	wg.Wait()
+
+	return c.firstErr()
 }
 
-func (c *Chain) Put(key string, val []byte) (err error) {
-	if !IsValidKey(key) {
+func (c *Chain) Put(ctx context.Context, key string, val []byte) (err error) {
+	if !isValidKey(key) {
 		return ErrInvalidKey
 	}
 
-	fn := func() (err error) {
-		for _, s := range c.stores {
-			err = s.Put(key, val)
-			if err != nil {
-				return
-			}
-		}
-		return
+	if err := c.firstErr(); err != nil {
+		return fmt.Errorf("Open failed due to a previous error: %q", err)
 	}
-	if c.async {
-		go fn()
-	} else {
-		err = fn()
+
+	fn := func(s Store) error {
+		return s.Put(ctx, key, val)
 	}
-	return
+
+	return c.doWithContext(ctx, fn)
 }
 
-func (c *Chain) Get(key string) (val []byte, err error) {
-	if !IsValidKey(key) {
+func (c *Chain) Get(ctx context.Context, key string) (val []byte, err error) {
+	if !isValidKey(key) {
 		return nil, ErrInvalidKey
 	}
 
-	for i, s := range c.stores {
-		val, err = s.Get(key)
-		if err != nil {
-			return
-		}
+	if err := c.firstErr(); err != nil {
+		return nil, fmt.Errorf("Open failed due to a previous error: %q", err)
+	}
 
-		if len(val) > 0 {
-			if i > 0 {
-				// put the value in all other stores up the chain
-				fn := func() {
-					for n := i - 1; n >= 0; n-- {
-						c.stores[n].Put(key, val) // errors..?
-					}
+	nextStore := make(chan Store, len(c.stores))
```

Comment:
Hey, I have a working snippet for you... without firstVal channel and the extra goroutine:

```go
package main

import (
	"errors"
	"fmt"
	"time"

	"golang.org/x/net/context"
)

func Get(ctx context.Context) (string, error) {
	nextStore := make(chan string, 3)
	for _, store := range []string{"first", "second", "third"} {
		nextStore <- store
	}
	close(nextStore)

	putBack := make(chan string, 3)

	for {
		select {
		case <-ctx.Done():
			return "", ctx.Err()
		case store, ok := <-nextStore:
			if !ok {
				return "", errors.New("not found")
			}

			if store != "fourth" { // not found
				putBack <- store
				break
			}
			close(putBack)

			for store := range putBack {
				go fmt.Printf("putting back to %v store\n", store)
			}

			return fmt.Sprintf("found in %v", store), nil
		}
	}

	panic("reached")
}

func main() {
	ctx, _ := context.WithTimeout(context.Background(), time.Second)
	val, err := Get(ctx)
	fmt.Println(val, err)

	// wait for fmt.Printf() from goroutines
	time.Sleep(time.Second)
}
```